mcp/src/test/java/io/modelcontextprotocol/client/McpAsyncClientResponseHandlerTests.java

```diff
@@ -27,6 +28,83 @@
 
 class McpAsyncClientResponseHandlerTests {
 
+	private InitializeResult initialization(McpAsyncClient asyncMcpClient, MockMcpTransport transport) {
+		// Create mock server response
+		McpSchema.InitializeResult mockInitResult = new McpSchema.InitializeResult(McpSchema.LATEST_PROTOCOL_VERSION,
+				McpSchema.ServerCapabilities.builder()
+					.tools(true)
+					.resources(true, true) // Enable both resources and resource templates
+					.build(),
+				new McpSchema.Implementation("test-server", "1.0.0"), "Test instructions");
+
+		// Create a Mono that will handle the initialization and response simulation
+		return asyncMcpClient.initialize().doOnSubscribe(subscription -> {
+			// Run in a separate reactive context to avoid blocking the main subscription
+			Mono.fromRunnable(() -> {
+				McpSchema.JSONRPCRequest initRequest = transport.getLastSentMessageAsRequest();
+				assertThat(initRequest.method()).isEqualTo(McpSchema.METHOD_INITIALIZE);
+
+				// Send mock server response
+				McpSchema.JSONRPCResponse initResponse = new McpSchema.JSONRPCResponse(McpSchema.JSONRPC_VERSION,
+						initRequest.id(), mockInitResult, null);
+				transport.simulateIncomingMessage(initResponse);
+			}).subscribeOn(reactor.core.scheduler.Schedulers.boundedElastic()).subscribe();
```

Comment:
This shouldn't be required. I think it's a source of the synchronization issues you are facing. Let me think about the way to address this.